packages/http-server-csharp/src/lib/service.ts

```diff
@@ -268,30 +272,120 @@ export async function $onEmit(context: EmitContext<CSharpServiceEmitterOptions>)
+      ${exceptionConstructor ? `${exceptionConstructor}\n` : ""}${this.emitter.emitModelProperties(model)}
     }
    } `,
       );
 
       return decl;
     }
 
+    getModelExceptionConstructor(program: Program, model: Model, className: string): string {
+      function getDefinedStatusCode() {
+        const statusCodeProperty = new ModelInfo().filterAllProperties(program, model, (p) =>
+          isStatusCode(program, p),
+        );
+
+        if (!statusCodeProperty) return undefined;
+
+        const { type } = statusCodeProperty;
+        switch (type.kind) {
+          case "Union": {
+            const firstVariant = type.variants.values().next().value;
+            return firstVariant?.type.kind === "Number" ? firstVariant.type.value : undefined;
+          }
+          case "Number":
+            return type.value;
+          default:
+            return getMinValue(program, statusCodeProperty) ?? undefined;
+        }
+      }
+
+      if (!isErrorModel(program, model)) {
+        return "";
+      }
+
+      const definedStatusCode = getDefinedStatusCode();
+      const constructor = this.getExceptionConstructorData(program, model);
+
+      return `public ${className}(${constructor.properties}) : base(${definedStatusCode ? definedStatusCode : "default"}${constructor.header ? `, \n\t\t headers: new(){${constructor.header}}` : ""}${constructor.value ? `, \n\t\t value: new{${constructor.value}}` : ""}) 
+        { ${constructor.body ? `\n${constructor.body}` : ""}\n\t}`;
+    }
+
+    getExceptionConstructorData(program: Program, model: Model) {
+      const allProperties = new ModelInfo().getAllProperties(program, model) ?? [];
+      const propertiesWithDefaults = allProperties.map((prop) => {
+        const { defaultValue: typeDefault } = this.#findPropertyType(prop);
+        const defaultValue = prop.defaultValue
+          ? code`${JSON.stringify(serializeValueAsJson(program, prop.defaultValue, prop))}`
+          : typeDefault;
+        return { prop, defaultValue };
+      });
+
+      const sortedProperties = propertiesWithDefaults
+        .filter(({ prop }) => !isStatusCode(program, prop))
+        .sort(({ prop: a, defaultValue: aDefault }, { prop: b, defaultValue: bDefault }) => {
+          if (!a.optional && !aDefault && (b.optional || bDefault)) return -1;
+          if (!b.optional && !bDefault && (a.optional || aDefault)) return 1;
+          return 0;
+        });
+
+      const properties: string[] = [];
+      const body: string[] = [];
+      const header: string[] = [];
+      const value: string[] = [];
+      for (const { prop, defaultValue } of sortedProperties) {
+        let propertyName = ensureCSharpIdentifier(program, prop, prop.name);
+        const modelName: string | undefined = this.emitter.getContext()["name"];
```

Comment:
It feels like we should pass the name in as a parameter, rather than getting it from the context